cpp/src/arrow/compute/kernels/scalar_compare.cc

```diff
@@ -746,6 +808,279 @@ std::shared_ptr<ScalarFunction> MakeScalarMinMax(std::string name,
+      switch (state.options.inclusive) {
+        case BetweenOptions::Inclusive::BOTH:
+          return GenerateDecimal<ScalarTernaryEqualTypes, BooleanType,
+                                 Op<BetweenOptions::Inclusive::BOTH>>(type_id)(ctx, batch,
+                                                                               out);
+        case BetweenOptions::Inclusive::LEFT:
+          return GenerateDecimal<ScalarTernaryEqualTypes, BooleanType,
+                                 Op<BetweenOptions::Inclusive::LEFT>>(type_id)(ctx, batch,
+                                                                               out);
+        case BetweenOptions::Inclusive::RIGHT:
+          return GenerateDecimal<ScalarTernaryEqualTypes, BooleanType,
+                                 Op<BetweenOptions::Inclusive::RIGHT>>(type_id)(
+              ctx, batch, out);
+        case BetweenOptions::Inclusive::NEITHER:
+          return GenerateDecimal<ScalarTernaryEqualTypes, BooleanType,
+                                 Op<BetweenOptions::Inclusive::NEITHER>>(type_id)(
+              ctx, batch, out);
+        default:
+          return Status::NotImplemented("between inclusiveness not implemented: ",
+                                        state.options.ToString());
+      }
+    };
+    InputType ty(type_id);
+    DCHECK_OK(func->AddKernel({ty, ty, ty}, boolean(), exec, BetweenState::Init));
+  }
+
+  // Add kernels for date types
+  for (const auto type_id : {Type::DATE32, Type::DATE64}) {
+    auto exec = [type_id](KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+      // Resolve generator based on options
+      const auto& state = static_cast<const BetweenState&>(*ctx->state());
+      switch (state.options.inclusive) {
+        case BetweenOptions::Inclusive::BOTH:
+          return GeneratePhysicalNumeric<ScalarTernaryEqualTypes, BooleanType,
+                                         Op<BetweenOptions::Inclusive::BOTH>>(type_id)(
+              ctx, batch, out);
+        case BetweenOptions::Inclusive::LEFT:
+          return GeneratePhysicalNumeric<ScalarTernaryEqualTypes, BooleanType,
+                                         Op<BetweenOptions::Inclusive::LEFT>>(type_id)(
+              ctx, batch, out);
+        case BetweenOptions::Inclusive::RIGHT:
+          return GeneratePhysicalNumeric<ScalarTernaryEqualTypes, BooleanType,
+                                         Op<BetweenOptions::Inclusive::RIGHT>>(type_id)(
+              ctx, batch, out);
+        case BetweenOptions::Inclusive::NEITHER:
+          return GeneratePhysicalNumeric<ScalarTernaryEqualTypes, BooleanType,
+                                         Op<BetweenOptions::Inclusive::NEITHER>>(type_id)(
+              ctx, batch, out);
+        default:
+          return Status::NotImplemented("between inclusiveness not implemented: ",
+                                        state.options.ToString());
+      }
+    };
```

Comment:
Something like this wouldn't work as-is, but it would if we refactored `GeneratePhysicalNumeric` etc. as functor classes:
```c++
template <template <typename...> class GenerateFunc>
ArrayKernelExec MakeBetweenArrayExec(Type::type type_id) {
  auto exec = [type_id](KernelContext* ctx, const ExecBatch& batch, Datum* out) {
    // Resolve generator based on options
    const auto& state = static_cast<const BetweenState&>(*ctx->state());
    switch (state.options.inclusive) {
      case BetweenOptions::Inclusive::BOTH:
        return GenerateFunc<ScalarTernaryEqualTypes, BooleanType,
                            Op<BetweenOptions::Inclusive::BOTH>>(type_id)(
            ctx, batch, out);
      case BetweenOptions::Inclusive::LEFT:
        return GenerateFunc<ScalarTernaryEqualTypes, BooleanType,
                            Op<BetweenOptions::Inclusive::LEFT>>(type_id)(
            ctx, batch, out);
      case BetweenOptions::Inclusive::RIGHT:
        return GenerateFunc<ScalarTernaryEqualTypes, BooleanType,
                            Op<BetweenOptions::Inclusive::RIGHT>>(type_id)(
            ctx, batch, out);
      case BetweenOptions::Inclusive::NEITHER:
        return GenerateFunc<ScalarTernaryEqualTypes, BooleanType,
                            Op<BetweenOptions::Inclusive::NEITHER>>(type_id)(
            ctx, batch, out);
      default:
        return Status::NotImplemented("between inclusiveness not implemented: ",
                                      state.options.ToString());
    }
  };
}
```

Reply:
Ok.

Reply:
Created a routine to minimize duplication within scalar_compare.cc. It uses only one template at the moment for between options and uses an if statement to choose the appropriate routine. Many more changes would be needed to use a template since the routines take different arguments, perhaps such changes are best done in another issue? One of the few routines that uses two templates is [Memoize](https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/cache_internal.h#L175).

Reply:
To be able to simplify/generalize as @pitrou proposed, we would need to make the following changes:
* Add [`GenerateFixedSizeBinary()`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/codegen_internal.h#L1307)
* Convert `GenerateXXX` functions into functors (so they can be used as template parameters, too)
* Add `GenerateFunc` template parameter to `MakeBetweenArrayExec`.

I agree that this could be done in a follow-up PR (maybe as part of [ARROW-14134](https://issues.apache.org/jira/browse/ARROW-14134)).

Reply:
Thanks. At the moment, tests for Decimal types with 3 scalars, 2 arrays and 1 scalar, and 1 array and 2 scalars are failing. Trying to determine why.

Reply:
@edponce Once those are done, happy for further refactoring in [ARROW-14134](https://issues.apache.org/jira/browse/ARROW-14134)